proposals/0000-updated-partial-type-sigs.rst

```diff
@@ -0,0 +1,497 @@
+
+   - In a type signature, a named wildcard behaves much as one does today, though its kind will be printed
+     in the diagnostic along with the other information. Just like today, a suggestion will be included
+     to enable ``-XPartialTypeSignatures``.
+
+   - In a type outside a type signature (e.g., in an instance declaration or a data constructor type),
+     a named wildcard will induce a diagnostic including the wildcard's kind and any information GHC
+     can figure out about its value.
+
+   - In an expression, a named wildcard will induce a diagnostic including the
+     wildcard's type and any information GHC can figure out about its value,
+     including suggested replacements. In this way, an expression named
+     wildcard will behave like holes have.
+
+3. In the absence of ``-XNamedWildCards``, the use of a type variable that begins with an underscore will induce a warning
+   that it looks like an attempt to write a named wildcard. This warning will be
+   controlled by ``-Wpossible-named-wildcards``, on by default.
+     
+4. Enabling ``-XPartialTypeSignatures`` is necessary in order for GHC to accept a program with
+   named wildcards in type signatures. These signatures must also be written using the new
+   separator ``::?`` instead of the typical ``::``. That is, we would now write ::
+
+     quux ::? _w -> Bool
+     quux x = not x
+
+   The new separator would be a loud indication that the signature is *partial*. It induces GHC
+   to use its partial-type-signature algorithm instead of its typical type-checking algorithm.
+
+   Partial type signatures would work with elisions, too, allowing ::
+
+     wurble ::? _ -> _
+     wurble x = not x
+
+   Named wildcards would induce diagnostics; elisions would not.
+
+   Partial type signatures would be kind-generalized *after* checking the function body. This would
+   allow something like the following to be accepted::
+
+     silly ::? Proxy a -> ()
+     silly (_ :: Proxy @Bool _) = ()
+
+   Note that the expression would be more specific than its type signature, if we kind-generalized
+   the signature *before* processing the expression.
+
+   This example is further explained below_.
+
+   Partial type signatures forbid polymorphic recursion, as they do today.
+
+5. Partial type signatures would generalize fresh variables only when an *extra-variables* wildcard
+   is in the type. That is, the last item in the list of variables after the word ``forall`` can
+   now be an elision ``_`` or a named wildcard ``_w`` (but only in a partial type signature).
+   In either case, this means that GHC can
+   generalize over more variables than have been written in the type signature. As usual, an elision
+   produces no diagnostic, while a named wildcard does. Here are two examples::
+
+     ex4 ::? _w -> _w
+     ex4 x = x
+
+     ex5 ::? forall _. _w -> _w
+     ex5 x = x
+
+   Here, ``ex4`` is rejected, because we do not know what type ``x`` should have and we cannot
```

Comment:
Because wildcards aren't something one can generalize over. Essentially, I'm saying here that you get to generalize over any out-of-scope variables in your signature -- and that's it. Instead, GHC currently generalizes over those variables and any variables it happens to need in order to fill in wildcards. This behavior is useful, but I don't think it should be the default any more than generalizing over extra constraints should be the default.

Reply:
But why can't one generalize over wildcards? I guess I'm not convinced by the distinction between an elision and a named wildcard in a type signature. Like @int-index, I feel like an elision should just be a named wildcard for some fresh name, and the main reason to use a named wildcard directly is to force a unification before handing things off to ghc.